### PR TITLE
Pluggable BuiltInExecutorBuilders

### PR DESCRIPTION
**Pull request by prdoyle**
Make the built-in executor builders pluggable so serverless can customize them.

See [ES-8974](https://elasticco.atlassian.net/browse/ES-8974).

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/node/NodeConstruction.java [hunk 1]
```diff
@@ -479,6 +481,7 @@ private ThreadPool createThreadPool(Settings settings, MeterRegistry meterRegist
         ThreadPool threadPool = new ThreadPool(
             settings,
             meterRegistry,
+            pluginsService.loadSingletonServiceProvider(BuiltInExecutorBuilders.class, DefaultBuiltInExecutorBuilders::new),
```
Comment (prdoyle) [inline server/src/main/java/org/elasticsearch/node/NodeConstruction.java:484]: The SPI magic 🪄

**Review by rjernst — COMMENTED** [review]
looks pretty good at a glance. one nit

File: server/src/main/java/org/elasticsearch/threadpool/Util.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.threadpool;
+
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.unit.ByteSizeValue;
+
+public class Util {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/threadpool/Util.java:14]: What do you think of having these methods on BuiltinExecutorBuilders? Alternatively, could this be a more concrete name like ThreadPoolUtil?

**Review by rjernst — APPROVED** [review]
Looks good, one suggestion

File: server/src/main/java/org/elasticsearch/threadpool/Util.java (hunk 2 above)
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/threadpool/Util.java:14]: ++ to have them in BuiltinExecutorBuilders

Comment (prdoyle) [reply server/src/main/java/org/elasticsearch/threadpool/Util.java:14]: Ok sure, let me try that. It has been a week and a half; I forget why I didn't do it that way in the first place.

Comment (prdoyle) [reply server/src/main/java/org/elasticsearch/threadpool/Util.java:14]: Alright, most of these methods are actually used exclusively by `DefaultBuildinExecutorBuilders` with the exception of two:
- `oneEighthAllocatedProcessors` is called by our old friend `Downsample`
- `searchOrGetThreadPoolSize` is called in a number of places:
  - `ESQL`
  - `QueryPragmas`
  - `TransportMultiSearchAction`

The ideal home for these methods depends, I think, on whether this is a pattern we want to encourage. Does it make sense for these callers to be calling these methods? If so, they probably belong someplace more "universal" than `BuiltinExecutorBuilders`.

Perhaps I could even put some or all of them back into `ThreadPool`; again, I forget why I moved them.

Comment (prdoyle) [reply server/src/main/java/org/elasticsearch/threadpool/Util.java]: Ok I moved the functions back where they were, in `ThreadPool`. That seems to work fine.

Comment (prdoyle) [reply server/src/main/java/org/elasticsearch/threadpool/Util.java]: I think in an earlier iteration of this code, I moved things into different packages, which made the utility methods inaccessible. But in the current code, they work just fine right where they are.

**Pull request merged**